## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#4909f2cd-eaaa-48b5-919e-7a647ad57bc1) |
| mint-large-bucket.sh | :x: [more...](#ee20ea28-a1b3-45e8-9c6b-afde5e02335f) |
| mint-dist-xl.sh | :x: [more...](#012a4a25-251a-4b2e-b109-081fa4a60c63) |
<a name="ee20ea28-a1b3-45e8-9c6b-afde5e02335f"></a>
### 9908-6a76044/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:32043
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 85f74864e144:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... FAILED in 12 minutes and 16 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-g5gn22ro04njkz",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 735858,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 161B6104038A41C9, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go DeleteBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-java tests ... done in 4 seconds
(2/15) Running aws-sdk-php tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-g5gn22ro04njkz",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 735858,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 161B6104038A41C9, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go DeleteBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(2/15) Running aws-sdk-ruby tests ... FAILED in 3 seconds
{
  "name": "aws-sdk-ruby",
  "function": "listBuckets",
  "args": {},
  "duration": 2183.3,
  "error": "Server not initialized, please try again.",
  "status": "FAIL"
}
(2/15) Running awscli tests ... FAILED in 14 seconds
{
  "name": "awscli",
  "duration": 12226,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (XMinioServerNotInitialized) when calling the CreateBucket operation (reached max retries: 4): Server not initialized, please try again."
}
(2/15) Running healthcheck tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": null,
  "duration": 129,
  "function": "testReadinessEndpoint",
  "message": "GET /minio/health/ready returned non OK status",
  "name": "healthcheck",
  "status": "FAIL"
}
(2/15) Running mc tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": null,
  "duration": 129,
  "function": "testReadinessEndpoint",
  "message": "GET /minio/health/ready returned non OK status",
  "name": "healthcheck",
  "status": "FAIL"
}
(2/15) Running minio-dotnet tests ... done in 23 seconds
(3/15) Running minio-go tests ... FAILED in 1 minutes and 14 seconds
{
  "args": {
    "bucketName": "minio-go-test-d5b9mv9kau416631",
    "region": "eu-west-1"
  },
  "duration": 73925,
  "error": "Server not initialized, please try again.",
  "function": "MakeBucket(bucketName, region)",
  "message": "MakeBucket failed",
  "name": "minio-go: testGetObjectClosedTwiceV2",
  "status": "FAIL"
}
(3/15) Running minio-java tests ... FAILED in 1 seconds
{
  "name": "minio-java",
  "function": "makeBucket(String bucketName)",
  "duration": 242,
  "status": "FAIL",
  "error": "java.lang.IllegalArgumentException: unknown error code string 'XMinioServerNotInitialized' >>> [io.minio.Xml.unmarshal(Xml.java:68), io.minio.MinioClient.execute(MinioClient.java:1134), io.minio.MinioClient.execute(MinioClient.java:1236), io.minio.MinioClient.executePut(MinioClient.java:1410), io.minio.MinioClient.makeBucket(MinioClient.java:3286), io.minio.MinioClient.makeBucket(MinioClient.java:3199), FunctionalTest.makeBucket_test1(FunctionalTest.java:247), FunctionalTest.runTests(FunctionalTest.java:4005), FunctionalTest.main(FunctionalTest.java:4192)]"
}
(3/15) Running minio-js tests ... FAILED in 0 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 5,
  "status": "FAIL",
  "error": "S3Error: Server not initialized, please try again. at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}
(3/15) Running minio-py tests ... FAILED in 10 minutes and 0 seconds
{
  "name": "minio-py:test_make_bucket_default_region",
  "function": "make_bucket(bucket_name, location)",
  "args": {
    "bucket_name": "minio-py-test-f0c24592-7ad2-4f58-b379-48abb2cb98e6",
    "location": "default value ('us-east-1')"
  },
  "duration": 600316,
  "message": "HTTPConnectionPool(host='minio-c3.minio.io', port=32043): Max retries exceeded with url: /minio-py-test-f0c24592-7ad2-4f58-b379-48abb2cb98e6/ (Caused by ResponseError('too many 503 error responses',))",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 188, in test_make_bucket_default_region\n    client.make_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 371, in make_bucket\n    headers=headers)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 336, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  [Previous line repeated 2 more times]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 805, in urlopen\n    retries = retries.increment(method, url, response=response, _pool=self)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 439, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minio-c3.minio.io', port=32043): Max retries exceeded with url: /minio-py-test-f0c24592-7ad2-4f58-b379-48abb2cb98e6/ (Caused by ResponseError('too many 503 error responses',))\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1968, in main\n    test_make_bucket_default_region(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 196, in test_make_bucket_default_region\n    raise Exception(err)\nException: HTTPConnectionPool(host='minio-c3.minio.io', port=32043): Max retries exceeded with url: /minio-py-test-f0c24592-7ad2-4f58-b379-48abb2cb98e6/ (Caused by ResponseError('too many 503 error responses',))\n",
  "status": "FAIL"
}
(3/15) Running s3cmd tests ... FAILED in 46 seconds
{
  "name": "s3cmd",
  "duration": "45456",
  "function": "test_make_bucket",
  "status": "FAIL",
  "error": "WARNING: Retrying failed request: / (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 3 sec...\nWARNING: Retrying failed request: / (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 6 sec...\nWARNING: Retrying failed request: / (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 9 sec...\nWARNING: Retrying failed request: / (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 12 sec...\nWARNING: Retrying failed request: / (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 15 sec...\nERROR: S3 error: 503 (XMinioServerNotInitialized): Server not initialized, please try again."
}
(3/15) Running s3select tests ... FAILED in 10 minutes and 1 seconds
{
  "name": "s3select:test_csv_input_quote_char",
  "function": "select_object_content(bucket_name, object_name, opts)",
  "args": {
    "bucket_name": "s3select-test-625e77d0-c621-4bfd-ac3c-b7a886a82b72"
  },
  "duration": 600517,
  "message": "HTTPConnectionPool(host='minio-c3.minio.io', port=32043): Max retries exceeded with url: /s3select-test-625e77d0-c621-4bfd-ac3c-b7a886a82b72/ (Caused by ResponseError('too many 503 error responses',))",
  "error": "Traceback (most recent call last):\n  File \"./tests.py\", line 45, in main\n    test_csv_input_custom_quote_char(client, log_output)\n  File \"/mint/run/core/s3select/csv.py\", line 82, in test_csv_input_custom_quote_char\n    client.make_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 371, in make_bucket\n    headers=headers)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 336, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  [Previous line repeated 2 more times]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 805, in urlopen\n    retries = retries.increment(method, url, response=response, _pool=self)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 439, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minio-c3.minio.io', port=32043): Max retries exceeded with url: /s3select-test-625e77d0-c621-4bfd-ac3c-b7a886a82b72/ (Caused by ResponseError('too many 503 error responses',))\n",
  "status": "FAIL"
}
(3/15) Running security tests ... done in 0 seconds

Executed 3 out of 15 tests successfully.
```
<a name="4909f2cd-eaaa-48b5-919e-7a647ad57bc1"></a>
### 9908-6a76044/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev3.minio.io:31091
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 2912be04a4e8:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... done in 2 minutes and 32 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 40 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 2 seconds
(9/15) Running minio-go tests ... FAILED in 51 seconds
{
  "args": {
    "bucketName": "minio-go-test-rp2kmgl2othuoa5l",
    "objectName": "6khoca4yt51oorkyo6u00lygxwc31q",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 2925,
  "error": "The bucket you tried to delete is not empty. You must delete all versions in the bucket.",
  "function": "RemoveObjectsWithOptions(bucketName, objectsCh, opts)",
  "message": "Cleanup failed",
  "name": "minio-go: testRemoveObjectsWithOptions",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... FAILED in 1 minutes and 48 seconds
{
  "name": "minio-java",
  "function": "enableObjectLegalHold(String bucketName, String objectName, String versionId)",
  "duration": 168,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = BucketNotEmpty, message = The bucket you tried to delete is not empty. You must delete all versions in the bucket., bucketName = minio-java-test-1uhatv9, objectName = null, resource = /minio-java-test-1uhatv9, requestId = 161B60C4157D7FE9, hostId = 1c93bba5-85f2-4a14-9d98-cc036e682dc5)\nrequest={method=DELETE, url=http://minio-dev3.minio.io:31091/minio-java-test-1uhatv9, headers=Host: minio-dev3.minio.io:31091\nAccept-Encoding: identity\nUser-Agent: MinIO (amd64; amd64) minio-java/dev\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20200624T043448Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20200624/us-east-1/s3/aws4_request, SignedHeaders=accept-encoding;content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=409, headers=Accept-Ranges: bytes\nContent-Length: 373\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO/DEVELOPMENT.2020-06-24T04-21-15Z\nVary: Origin\nX-Amz-Request-Id: 161B60C4157D7FE9\nX-Xss-Protection: 1; mode=block\nDate: Wed, 24 Jun 2020 04:34:48 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:1211), io.minio.MinioClient.execute(MinioClient.java:1236), io.minio.MinioClient.executeDelete(MinioClient.java:1366), io.minio.MinioClient.removeBucket(MinioClient.java:3686), FunctionalTest.enableObjectLegalHold_test(FunctionalTest.java:3314), FunctionalTest.runTests(FunctionalTest.java:4079), FunctionalTest.main(FunctionalTest.java:4192)]"
}
(9/15) Running minio-js tests ... done in 46 seconds
(10/15) Running minio-py tests ... done in 2 minutes and 22 seconds
(11/15) Running s3cmd tests ... done in 18 seconds
(12/15) Running s3select tests ... done in 3 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
<a name="012a4a25-251a-4b2e-b109-081fa4a60c63"></a>
### 9908-6a76044/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:30242
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp d11e64519d65:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... FAILED in 6 minutes and 15 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-kvyyinovovtvx2",
    "expiry": 60000000000,
    "objectName1": "testObject1",
    "objectName2": "testObject2"
  },
  "duration": 252166,
  "error": "BucketNotEmpty: The bucket you tried to delete is not empty\n\tstatus code: 409, request id: 161B60B020772C46, host id: ",
  "function": "testListObjects",
  "message": "AWS SDK Go DeleteBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-java tests ... done in 5 seconds
(2/15) Running aws-sdk-php tests ... done in 1 minutes and 0 seconds
(3/15) Running aws-sdk-ruby tests ... FAILED in 7 seconds
{
  "name": "aws-sdk-ruby",
  "function": "removeBucket(bucket_name)",
  "args": {
    "bucket_name": "aws-sdk-ruby-bucket-a5c296cfe87d"
  },
  "duration": 3552.18,
  "error": "Failed to clean-up bucket 'aws-sdk-ruby-bucket-a5c296cfe87d', The bucket you tried to delete is not empty",
  "status": "FAIL"
}
(3/15) Running awscli tests ... FAILED in 20 seconds
{
  "name": "awscli",
  "duration": 13877,
  "function": "delete_bucket\n",
  "status": "FAIL",
  "error": "delete: s3://awscli-mint-test-bucket-9747/datafile-1-kBremove_bucket failed: s3://awscli-mint-test-bucket-9747 An error occurred (BucketNotEmpty) when calling the DeleteBucket operation: The bucket you tried to delete is not empty"
}
(3/15) Running healthcheck tests ... done in 0 seconds
(4/15) Running mc tests ... done in 1 minutes and 8 seconds
(5/15) Running minio-dotnet tests ... done in 1 minutes and 31 seconds
(6/15) Running minio-go tests ... FAILED in 4 seconds
{
  "args": {
    "bucketName": "minio-go-test-cqsqtmprjpwoa6gg",
    "objectName": "fjvesty0910x66p0vqtcif130kwpjb",
    "region": "eu-west-1"
  },
  "duration": 3787,
  "error": "The bucket you tried to delete is not empty",
  "function": "MakeBucket(bucketName, region)",
  "message": "Cleanup failed",
  "name": "minio-go: testGetObjectClosedTwiceV2",
  "status": "FAIL"
}
(6/15) Running minio-java tests ... FAILED in 3 seconds
{
  "name": "minio-java",
  "function": "listBuckets()",
  "duration": 30,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = SlowDown, message = Please reduce your request, bucketName = minio-java-test-l4f0l1, objectName = null, resource = /minio-java-test-l4f0l1, requestId = 161B60EC1EAAB50C, hostId = 5e8880a1-1dc1-4c9a-b542-1f04ef5bcd1d)\nrequest={method=PUT, url=http://minio-c3.minio.io:30242/minio-java-test-l4f0l1, headers=Host: minio-c3.minio.io:30242\nAccept-Encoding: identity\nUser-Agent: MinIO (amd64; amd64) minio-java/dev\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20200624T043740Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20200624/us-east-1/s3/aws4_request, SignedHeaders=accept-encoding;content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=503, headers=Accept-Ranges: bytes\nContent-Length: 303\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nRetry-After: 120\nServer: MinIO/DEVELOPMENT.2020-06-24T04-21-15Z\nVary: Origin\nX-Amz-Request-Id: 161B60EC1EAAB50C\nX-Xss-Protection: 1; mode=block\nDate: Wed, 24 Jun 2020 04:37:40 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:1211), io.minio.MinioClient.execute(MinioClient.java:1236), io.minio.MinioClient.executePut(MinioClient.java:1410), io.minio.MinioClient.makeBucket(MinioClient.java:3286), io.minio.MinioClient.makeBucket(MinioClient.java:3199), FunctionalTest.listBuckets_test(FunctionalTest.java:324), FunctionalTest.runTests(FunctionalTest.java:4011), FunctionalTest.main(FunctionalTest.java:4192)]"
}
(6/15) Running minio-js tests ... FAILED in 6 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 7,
  "status": "FAIL",
  "error": "S3Error: Please reduce your request at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}
(6/15) Running minio-py tests ... FAILED in 6 minutes and 3 seconds
{
  "name": "minio-py:test_fput_object_small_file",
  "function": "fput_object(bucket_name, object_name, file_path, content_type, metadata, sse, progress, part_size)",
  "args": {
    "bucket_name": "minio-py-test-8355648a-203e-4ee0-bccb-824018207a33",
    "object_name": "7174f53d-0b81-483c-acaa-a875876a6996",
    "file_path": "/mint/data/datafile-1-MB",
    "metadata": {
      "x-amz-storage-class": "STANDARD_IA"
    }
  },
  "duration": 240457,
  "message": "BucketNotEmpty: message: The bucket you tried to delete is not empty.",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 371, in test_fput_object_small_file\n    client.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 446, in remove_bucket\n    self._url_open('DELETE', bucket_name=bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 2020, in _url_open\n    object_name).get_exception()\nminio.error.BucketNotEmpty: BucketNotEmpty: message: The bucket you tried to delete is not empty.\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1983, in main\n    test_fput_object_small_file(client, testfile, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 373, in test_fput_object_small_file\n    raise Exception(err)\nException: BucketNotEmpty: message: The bucket you tried to delete is not empty.\n",
  "status": "FAIL"
}
(6/15) Running s3cmd tests ... done in 25 seconds
(7/15) Running s3select tests ... FAILED in 40 minutes and 5 seconds
{
  "name": "s3select:test_csv_input_quote_char",
  "function": "select_object_content(bucket_name, object_name, opts)",
  "args": {
    "bucket_name": "s3select-test-1a5f3e4f-3547-4fba-b028-a007d5a1fcd1"
  },
  "duration": 2404189,
  "message": "BucketNotEmpty: message: The bucket you tried to delete is not empty.",
  "error": "Traceback (most recent call last):\n  File \"./tests.py\", line 45, in main\n    test_csv_input_custom_quote_char(client, log_output)\n  File \"/mint/run/core/s3select/csv.py\", line 110, in test_csv_input_custom_quote_char\n    client.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 446, in remove_bucket\n    self._url_open('DELETE', bucket_name=bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 2020, in _url_open\n    object_name).get_exception()\nminio.error.BucketNotEmpty: BucketNotEmpty: message: The bucket you tried to delete is not empty.\n",
  "status": "FAIL"
}
(7/15) Running security tests ... done in 0 seconds

Executed 7 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/config/cache/config.go
+++ b/cmd/config/cache/config.go
@@ -37,6 +37,7 @@ type Config struct {
 	After         int      `json:"after"`
 	WatermarkLow  int      `json:"watermark_low"`
 	WatermarkHigh int      `json:"watermark_high"`
+	Range         bool     `json:"range"`
 }
 
 // UnmarshalJSON - implements JSON unmarshal interface for unmarshalling

--- a/cmd/config/cache/help.go
+++ b/cmd/config/cache/help.go
@@ -68,5 +68,11 @@ var (
 			Optional:    true,
 			Type:        "number",
 		},
+		config.HelpKV{
+			Key:         Range,
+			Description: `set to "on" or "off" caching of independent range requests per object, defaults to "on"`,
+			Optional:    true,
+			Type:        "string",
+		},
 	}
 )

--- a/cmd/config/cache/lookup.go
+++ b/cmd/config/cache/lookup.go
@@ -34,6 +34,7 @@ const (
 	After         = "after"
 	WatermarkLow  = "watermark_low"
 	WatermarkHigh = "watermark_high"
+	Range         = "range"
 
 	EnvCacheDrives        = "MINIO_CACHE_DRIVES"
 	EnvCacheExclude       = "MINIO_CACHE_EXCLUDE"
@@ -43,6 +44,7 @@ const (
 	EnvCacheAfter         = "MINIO_CACHE_AFTER"
 	EnvCacheWatermarkLow  = "MINIO_CACHE_WATERMARK_LOW"
 	EnvCacheWatermarkHigh = "MINIO_CACHE_WATERMARK_HIGH"
+	EnvCacheRange         = "MINIO_CACHE_RANGE"
 
 	EnvCacheEncryptionMasterKey = "MINIO_CACHE_ENCRYPTION_MASTER_KEY"
 
@@ -83,6 +85,10 @@ var (
 		config.KV{
 			Key:   WatermarkHigh,
 			Value: DefaultWaterMarkHigh,
+		},
+		config.KV{
+			Key:   Range,
+			Value: config.EnableOn,
 		},
 	}
 )
@@ -195,5 +201,15 @@ func LookupConfig(kvs config.KVS) (Config, error) {
 		err := errors.New("config high watermark value should be greater than low watermark value")
 		return cfg, config.ErrInvalidCacheWatermarkHigh(err)
 	}
+
+	cfg.Range = true // by default range caching is enabled.
+	if rangeStr := env.Get(EnvCacheRange, kvs.Get(Range)); rangeStr != "" {
+		rng, err := config.ParseBool(rangeStr)
+		if err != nil {
+			return cfg, config.ErrInvalidCacheRange(err)
+		}
+		cfg.Range = rng
+	}
+
 	return cfg, nil
 }

--- a/cmd/config/errors.go
+++ b/cmd/config/errors.go
@@ -96,6 +96,12 @@ var (
 		"MINIO_CACHE_ENCRYPTION_MASTER_KEY: For more information, please refer to https://docs.min.io/docs/minio-disk-cache-guide",
 	)
 
+	ErrInvalidCacheRange = newErrFn(
+		"Invalid cache range value",
+		"Please check the passed value",
+		"MINIO_CACHE_RANGE: Valid expected value is `on` or `off`",
+	)
+
 	ErrInvalidRotatingCredentialsBackendEncrypted = newErrFn(
 		"Invalid rotating credentials",
 		"Please set correct rotating credentials in the environment for decryption",

--- a/cmd/disk-cache-backend.go
+++ b/cmd/disk-cache-backend.go
@@ -33,6 +33,7 @@ import (
 	"time"
 
 	"github.com/djherbis/atime"
+	"github.com/minio/minio/cmd/config/cache"
 	"github.com/minio/minio/cmd/crypto"
 	xhttp "github.com/minio/minio/cmd/http"
 	"github.com/minio/minio/cmd/logger"
@@ -137,6 +138,7 @@ type diskCache struct {
 	after         int // minimum accesses before an object is cached.
 	lowWatermark  int
 	highWatermark int
+	enableRange   bool
 	// nsMutex namespace lock
 	nsMutex *nsLockMap
 	// Object functions pointing to the corresponding functions of backend implementation.
@@ -144,7 +146,12 @@ type diskCache struct {
 }
 
 // Inits the disk cache dir if it is not initialized already.
-func newDiskCache(ctx context.Context, dir string, quotaPct, after, lowWatermark, highWatermark int) (*diskCache, error) {
+func newDiskCache(ctx context.Context, dir string, config cache.Config) (*diskCache, error) {
+	quotaPct := config.MaxUse
+	if quotaPct == 0 {
+		quotaPct = config.Quota
+	}
+
 	if err := os.MkdirAll(dir, 0777); err != nil {
 		return nil, fmt.Errorf("Unable to initialize '%s' dir, %w", dir, err)
 	}
@@ -153,9 +160,10 @@ func newDiskCache(ctx context.Context, dir string, quotaPct, after, lowWatermark
 		triggerGC:     make(chan struct{}),
 		stats:         CacheDiskStats{Dir: dir},
 		quotaPct:      quotaPct,
-		after:         after,
-		lowWatermark:  lowWatermark,
-		highWatermark: highWatermark,
+		after:         config.After,
+		lowWatermark:  config.WatermarkLow,
+		highWatermark: config.WatermarkHigh,
+		enableRange:   config.Range,
 		online:        1,
 		pool: sync.Pool{
 			New: func() interface{} {

--- a/cmd/disk-cache.go
+++ b/cmd/disk-cache.go
@@ -312,6 +312,10 @@ func (c *cacheObjects) GetObjectNInfo(ctx context.Context, bucket, object string
 
 	if rs != nil {
 		go func() {
+			// if range caching is disabled, download entire object.
+			if !dcache.enableRange {
+				rs = nil
+			}
 			// fill cache in the background for range GET requests
 			bReader, bErr := c.GetObjectNInfoFn(ctx, bucket, object, rs, h, lockType, opts)
 			if bErr != nil {
@@ -321,7 +325,9 @@ func (c *cacheObjects) GetObjectNInfo(ctx context.Context, bucket, object string
 			oi, _, _, err := dcache.statRange(ctx, bucket, object, rs)
 			// avoid cache overwrite if another background routine filled cache
 			if err != nil || oi.ETag != bReader.ObjInfo.ETag {
-				dcache.Put(ctx, bucket, object, bReader, bReader.ObjInfo.Size, rs, ObjectOptions{UserDefined: getMetadata(bReader.ObjInfo)}, false)
+				// use a new context to avoid locker prematurely timing out operation when the GetObjectNInfo returns.
+				dcache.Put(context.Background(), bucket, object, bReader, bReader.ObjInfo.Size, rs, ObjectOptions{UserDefined: getMetadata(bReader.ObjInfo)}, false)
+				return
 			}
 		}()
 		return bkReader, bkErr
@@ -542,11 +548,7 @@ func newCache(config cache.Config) ([]*diskCache, bool, error) {
 			return nil, false, errors.New("Atime support required for disk caching")
 		}
 
-		quota := config.MaxUse
-		if quota == 0 {
-			quota = config.Quota
-		}
-		cache, err := newDiskCache(ctx, dir, quota, config.After, config.WatermarkLow, config.WatermarkHigh)
+		cache, err := newDiskCache(ctx, dir, config)
 		if err != nil {
 			return nil, false, err
 		}

--- a/docs/disk-caching/DESIGN.md
+++ b/docs/disk-caching/DESIGN.md
@@ -15,6 +15,7 @@ minio gateway <name> -h
      MINIO_CACHE_AFTER: Minimum number of access before caching an object.
      MINIO_CACHE_WATERMARK_LOW: % of cache quota at which cache eviction stops
      MINIO_CACHE_WATERMARK_HIGH: % of cache quota at which cache eviction starts
+     MINIO_CACHE_RANGE: set to "on" or "off" caching of independent range requests per object, defaults to "on"
 
 
 ...
@@ -62,6 +63,7 @@ Disk caching caches objects for **downloaded** objects i.e
 - When an object is deleted, corresponding entry in cache if any is deleted as well.
 - Cache continues to work for read-only operations such as GET, HEAD when backend is offline.
 - Cache-Control and Expires headers can be used to control how long objects stay in the cache. ETag of cached objects are not validated with backend until expiry time as per the Cache-Control or Expires header is met.
+- All range GET requests are cached by default independently, this may be not desirable in all situations when cache storage is limited and where downloading an entire object at once might be more optimal. To optionally turn this feature off, and allow downloading entire object in the background `export MINIO_CACHE_RANGE=off`.
 - To ensure security guarantees, encrypted objects are normally not cached. However, if you wish to encrypt cached content on disk, you can set MINIO_CACHE_ENCRYPTION_MASTER_KEY environment variable to set a cache KMS
 master key to automatically encrypt all cached content.
 

